hocuspocus-server/src/rejectUnauthorized.ts

```diff
@@ -0,0 +1,130 @@
+    /**
+     * The messages we are interested in are of the following format:
+     * [docIdLength: number, ...docIdString: string, hocuspocusMessageType: number,  ySyncMessageType: number, ...yjsUpdate: Uint8Array]
+     *
+     * We check that the hocuspocusMessageType is Sync and that the ySyncMessageType is messageYjsUpdate.
+     */
+    const incomingMessage = new IncomingMessage(message);
+    // Read the docID string, but don't use it
+    incomingMessage.readVarString();
+
+    // Read the hocuspocusMessageType
+    const hocuspocusMessageType = incomingMessage.readVarUint();
+    // If the hocuspocusMessageType is not Sync, we don't handle the message, since it is not an update
+    if (hocuspocusMessageType !== MessageType.Sync) {
+      return;
+    }
+
+    // Read the ySyncMessageType
+    const ySyncMessageType = incomingMessage.readVarUint();
+
+    // If the ySyncMessageType is not messageYjsUpdate, we don't handle the message, since it is not an update
+    if (ySyncMessageType !== syncProtocol.messageYjsUpdate) {
+      // not an update
+      return;
+    }
+
+    // Read the yjsUpdate
+    const yUpdate = incomingMessage.readVarUint8Array();
+
+    return yUpdate;
+  }
+
+  /**
+   * This function protects against changes to the restricted type.
+   * It does this by:
+   * - setting up an undo manager on the restricted type
+   * - caching pending updates from the Ydoc to avoid certain attacks
+   * - applying the received update and checking whether the restricted type has been changed
+   * - catching errors that might try to circumvent the restrictions
+   * - undoing changes on restricted types
+   * - reapplying pending updates
+   *
+   * @param yUpdate The update to apply
+   * @param ydoc The document that the update is being applied to
+   * @param restrictedType The type that we want to protect
+   * @returns true if the update was rejected, false otherwise
+   */
+  private rollbackUpdateIfNeeded(
+    yUpdate: Uint8Array,
+    ydoc: Y.Doc,
+    restrictedType: Y.AbstractType<any>
+  ) {
+    // don't handle changes of the local undo manager, which is used to undo invalid changes
+    const um = new Y.UndoManager(restrictedType, {
+      trackedOrigins: new Set(["remote change"]),
+    });
+    const beforePendingDs = ydoc.store.pendingDs;
+    const beforePendingStructs = ydoc.store.pendingStructs?.update;
+    let didNeedToUndo = false;
+    try {
+      Y.applyUpdate(ydoc, yUpdate, "remote change");
+    } finally {
+      while (um.undoStack.length) {
+        um.undo();
+        didNeedToUndo = true;
+      }
+      um.destroy();
+      ydoc.store.pendingDs = beforePendingDs;
+      ydoc.store.pendingStructs = null;
+      if (beforePendingStructs) {
+        Y.applyUpdateV2(ydoc, beforePendingStructs);
+      }
+    }
+
+    return didNeedToUndo;
+  }
+
+  async beforeHandleMessage({
+    update,
+    document: ydoc,
+  }: beforeHandleMessagePayload) {
+    const yUpdate = this.getYUpdate(update);
```

Comment:
I see we process the update here (and in `rollbackUpdateIfNeeded` manually). Wouldn't we now apply the update twice? i.e.: the regular hocuspocus code will still process the message, right? so everything would be processed twice?

Reply:
The only way that I could think of not applying the update twice would be to make a whole copy of the ydoc.

Applying updates twice should be fine, no? It is a CRDT 😄

Reply:
to be clear; this means all updates will be applied twice. Wouldn't that double the load?